src/nginx_manager.py

```diff
@@ -3,73 +3,238 @@
+        # The default permission for nginx configuration files are 755.
+        NGINX_SITES_AVAILABLE_PATH.mkdir(mode=0o755, parents=True, exist_ok=True)
+        NGINX_SITES_ENABLED_PATH.mkdir(mode=0o755, parents=True, exist_ok=True)
+        NGINX_SITES_AVAILABLE_PATH.chmod(mode=0o755)
+        NGINX_SITES_ENABLED_PATH.chmod(mode=0o755)
+    except (PermissionError, OSError, IOError) as err:
+        logger.exception("Failed to reset the sites configurations directories.")
+        raise NginxFileError("Failed to reset sites configurations") from err
+
+
+def _create_server_config(host: str, configuration: HostConfig) -> None:
+    """Create the nginx configuration file for a virtual host.
+
+    Args:
+        host: The name of the virtual host.
+        configuration: The configurations of the host.
+
+    Raises:
+        NginxConfigurationError: Failed to convert the configuration to nginx format.
+        NginxFileError: File operation errors while updating nginx configuration files.
+    """
+    logger.info("Creating the nginx site configuration file for hosts %s", host)
+    try:
         nginx_config = nginx.Conf()
-        server_config = nginx.Server()
-        for i, config in enumerate(configuration.values()):
-            backend_set_name = f"backend_{i}"
+        server_config = nginx.Server(
+            nginx.Key("server_name", host),
+            nginx.Key("access_log", _get_access_log_path(host)),
+            nginx.Key("error_log", _get_error_log_path(host)),
+        )
 
-            hosts = [nginx.Key("server", host) for host in config.backends]
+        for path, config in configuration.items():
+            host_with_path = host + path.rstrip("/")
 
-            upstream_config = nginx.Upstream(backend_set_name, *hosts)
+            backends = [nginx.Key("server", ip) for ip in config.backends]
+            upstream_config = nginx.Upstream(host_with_path, *backends)
             nginx_config.add(upstream_config)
             server_config.add(
                 nginx.Location(
-                    config.location,
-                    nginx.Key("proxy_pass", f"{config.protocol}://{backend_set_name}"),
+                    path,
+                    nginx.Key("proxy_pass", f"{config.protocol.value}://{host_with_path}"),
+                    nginx.Key("proxy_set_header", f'Host "{host}"'),
                 )
             )
+
         nginx_config.add(server_config)
-        nginx.dumpf(nginx_config, CHARM_MANAGED_NGINX_CONFIG_PATH)
+    except nginx.ParseError as err:
+        logger.exception(
+            "Unable to convert %s configuration to nginx format: %s", host, configuration
+        )
+        raise NginxConfigurationError(
+            f"Unable to convert {host} configuration to nginx format: {configuration}"
+        ) from err
+
+    try:
+        nginx.dumpf(nginx_config, _get_sites_available_path(host))
+        _get_sites_enabled_path(host).symlink_to(_get_sites_available_path(host))
```

Comment:
Perhaps add a small comment to explain why you're doing this, somebody less familiar with nginx might not understand

Reply:
Also both _get_sites_enabled_path and _get_sites_available_path are only used here and are 1 liner methods
I'd rather have a method called `enable_site(host: str)`
That will build the path, and create the symlink, then no need for a comment because the method will be documented via docstrings

Reply:
I added a function that creates the nginx config file and add symlink, with the docstring explaining why.

I did tried to delete the `_get_sites_enabled_path` and `_get_sites_available_path` functions, but I think it might be better to keep them as they are used in tests. It is likely better for the test to use the same function to get the path as the charm to reduce code duplicate and regressions.